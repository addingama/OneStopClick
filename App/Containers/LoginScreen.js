import React, { Component, PropTypes } from 'react'
import { View, Image, ScrollView, Text } from 'react-native'
import I18n from 'react-native-i18n'
import { NavigationActions } from 'react-navigation'
import { connect } from 'react-redux'
import ProgressIndicator from '../Components/ProgressIndicator'
import LoginActions from '../Redux/LoginRedux'
import { Images } from '../Themes'
import { OscInputField, OscButton } from '../Lib/formGenerator'
import * as LoginModel from '../Models/LoginModel'
import { validateField } from '../Lib/validator'

import styles from './Styles/LoginScreenStyle'

class LoginScreen extends Component {

  static propTypes = {
    dispatch: PropTypes.func,
    loggingIn: PropTypes.bool,
    error: PropTypes.bool,
    message: PropTypes.string,
    access_token: PropTypes.string,
    refresh_token: PropTypes.string,
    attemptLogin: PropTypes.func.isRequired
  }

  constructor (props) {
    super(props)
    this.state = {
      fields: LoginModel.login
    }
    this.updateState = this.updateState.bind(this)
    this.handlePressLogin = this.handlePressLogin.bind(this)
  }

  updateState (newFieldState) {
    this.setState({fields: newFieldState})
  }

  validateFields () {
    var state = Object.assign({}, this.state)
    var isValid = true
    Object.keys(state.fields).map((field) => {
      const result = validateField(state.fields, field)
      this.updateState(result)
      if (!result[field].valid) {
        isValid = false
      }
    })
    return isValid
  }

  handlePressLogin () {
    if (this.validateFields()) {
      const { email, password } = this.state.fields
      console.tron.log(email, password)
      this.props.attemptLogin(email.value, password.value)
    }
  }

  goToRegistrationScreen () {
    const resetAction = NavigationActions.reset({
      index: 0,
      actions: [
        NavigationActions.navigate({ routeName: 'RegistrationScreen' })
      ]
    })
    this.props.navigation.dispatch(resetAction)
  }

  render () {
    const {email, password} = Object.assign({}, this.state.fields)
    const { loggingIn, error } = this.props
    return (
      <View style={styles.mainContainer}>
        <Image source={Images.background} style={styles.backgroundImage} resizeMode='stretch' />
        <ScrollView contentContainerStyle={{flexGrow: 1}}>
          <View style={styles.container}>
            <View style={styles.formContainer}>
              <Text style={styles.titleLoginText}>Login</Text>
              <OscInputField
                field={email}
                editable={!loggingIn}
                state={Object.assign({}, this.state.fields)}
                updateState={this.updateState}
              />
              <OscInputField
                field={password}
                editable={!loggingIn}
                state={Object.assign({}, this.state.fields)}
                updateState={this.updateState}
              />

              <OscButton
                disabled={loggingIn}
                onPress={() => this.handlePressLogin()}
                style={styles.btnSignIn}
                title={I18n.t('signIn')}
              />

              <View style={styles.doNotHaveAccount}>
                <Text>Do not have an account?</Text>
                <Text style={[styles.registerText]} onPress={() => this.goToRegistrationScreen()}> Register!</Text>
              </View>
            </View>
            <ProgressIndicator show={loggingIn} text='Loggin you in' />
          </View>
        </ScrollView>
      </View>

    )
  }
}

const mapStateToProps = (state) => {
  return {
    loggingIn: state.login.loggingIn,
    error: state.login.error,
    message: state.login.message,
    access_token: state.login.access_token,
    refresh_token: state.login.refresh_token
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    attemptLogin: (username, password) => dispatch(LoginActions.loginRequest(username, password))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(LoginScreen)
