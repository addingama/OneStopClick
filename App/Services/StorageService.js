import { AsyncStorage } from 'react-native'

const ACCESS_TOKEN = 'accessToken'
// const REFRESH_TOKEN = 'refreshToken'

class StorageService {
  static saveSession (accessToken) {
    AsyncStorage.setItem(ACCESS_TOKEN, accessToken)
    // AsyncStorage.setItem(REFRESH_TOKEN, refreshToken)
  }

  static removeSession () {
    AsyncStorage.removeItem(ACCESS_TOKEN)
    // AsyncStorage.removeItem(REFRESH_TOKEN)
  }

  static getAccessToken () {
    return AsyncStorage.getItem(ACCESS_TOKEN)
  }

  static isLoggedIn () {
    return this.getAccessToken().then((accessToken) => {
      return accessToken
    })
  }
}

export default StorageService
