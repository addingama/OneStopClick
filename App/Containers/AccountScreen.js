import React, { PropTypes } from 'react'
import { ScrollView, View, Alert, TouchableOpacity } from 'react-native'
import { Text } from 'react-native-elements'
import { connect } from 'react-redux'
import I18n from 'react-native-i18n'
import { NavigationActions } from 'react-navigation'
import StorageService from '../Services/StorageService'
import AccountDrawerBase from './Bases/AccountDrawerBase'
import DrawerHeader from '../Components/DrawerHeader'
import UserActions from '../Redux/UserRedux'
import LoginActions from '../Redux/LoginRedux'

// Styles
import styles from './Styles/AccountScreenStyle'

class AccountScreen extends AccountDrawerBase {
  static navigationOptions = AccountDrawerBase.getNavigationOptions()

  static propTypes = {
    dispatch: PropTypes.func,
    fetching: PropTypes.bool,
    error: PropTypes.bool,
    message: PropTypes.string,
    accessToken: PropTypes.string,
    user: PropTypes.object,
    attempGetProfile: PropTypes.func.isRequired,
    resetLogin: PropTypes.func,
    resetUser: PropTypes.func
  }

  componentWillMount () {
    if (this.props.accessToken == null) {
      return this.goToLogin()
    }
  }

  componentWillReceiveProps (newProps) {
    this.forceUpdate()

    if (!newProps.fetching && newProps.user == null && !newProps.error && newProps.accessToken != null) {
      this.props.attempGetProfile(newProps.accessToken)
    }
  }

  handlePressEdit () {
    this.props.navigation.navigate('EditProfileScreen')
  }

  handlePresslogout () {
    Alert.alert(
      I18n.t('message'),
      I18n.t('logoutConfirmation'),
      [
        { text: I18n.t('cancel') },
        { text: I18n.t('ok'), onPress: () => this.logout() }
      ],
      { cancelable: false }
    )
  }

  goToLogin () {
    const resetAction = NavigationActions.reset({
      index: 0,
      actions: [
        NavigationActions.navigate({
          routeName: 'LoginScreen',
          params: {}
        })
      ]
    })
    this.props.navigation.dispatch(resetAction)
  }

  logout () {
    // clear data here
    const resetAction = NavigationActions.navigate({
      routeName: 'Home',
      params: {},
      action: NavigationActions.navigate({ routeName: 'Home' })
    })
    StorageService.removeSession()
    // dispatch reset login and user state
    this.props.resetLogin()
    this.props.resetUser()
    this.props.navigation.dispatch(resetAction)
  }

  render () {
    const { user } = this.props
    return (
      <View>
        <View style={styles.hasNavbar}>
          <DrawerHeader title={I18n.t('account')} {...this.props} />
        </View>
        <ScrollView>
          <View>
            <View style={[styles.whiteSection, {flex: 1, flexDirection: 'row'}]}>
              <View>
                <Text style={styles.boldMediumText}>{user == null ? '' : user.name}</Text>
                <Text style={styles.smallText}>{user == null ? '' : user.email}</Text>
              </View>
              <View style={{ flex: 1, alignItems: 'flex-end' }}>
                <TouchableOpacity onPress={() => this.handlePressEdit()}>
                  <Text style={styles.editButton}>Edit</Text>
                </TouchableOpacity>
              </View>
            </View>

            <View style={[styles.whiteSection, {flex: 1, alignItems: 'center'}]}>
              <TouchableOpacity onPress={() => this.handlePresslogout()}>
                <Text style={styles.logoutButton}>{I18n.t('logOut')}</Text>
              </TouchableOpacity>
            </View>
          </View>
        </ScrollView>
      </View>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    fetching: state.user.fetching,
    error: state.user.error,
    message: state.user.message,
    accessToken: state.login.accessToken,
    user: state.user.user
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    attempGetProfile: (accessToken) => dispatch(UserActions.userProfileRequest(accessToken)),
    resetLogin: () => dispatch(LoginActions.logout()),
    resetUser: () => dispatch(UserActions.userReset())
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(AccountScreen)
