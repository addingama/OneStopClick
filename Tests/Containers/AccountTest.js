//__tests__/AccountTest.js
import 'react-native'
import React from 'react'
import AccountScreen from '../../App/Containers/AccountScreen'
import LoginScreen from '../../App/Containers/LoginScreen'
import StorageService from '../../App/Services/StorageService'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import createStore from '../../App/Redux'

const middlewares = []
const mockStore = configureStore(middlewares)

describe('Account Screen', () => {
    test('renderer correctly', () => {
    const store = mockStore()
    const tree = renderer.create(
    <Provider store={store}>
         <AccountScreen />
      </Provider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
    })
})