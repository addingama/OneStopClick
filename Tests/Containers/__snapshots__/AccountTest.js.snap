// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Screen renderer correctly 1`] = `
<View>
  <View
    style={
      Object {
        "marginBottom": 84,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
            "borderBottomColor": "#f2f2f2",
            "borderBottomWidth": 1,
            "height": 70,
            "left": 0,
            "padding": 15,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "backgroundColor": "#2F1F37",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "flex-end",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            },
            undefined,
          ]
        }
      >
        <View>
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hitSlop={undefined}
            isTVSelectable={true}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginLeft": 9,
                "marginTop": 7,
                "opacity": 1,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <View
              onPress={undefined}
              style={
                Array [
                  false,
                  false,
                  false,
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
              underlayColor="white"
            >
              <Text
                accessible={true}
                allowFontScaling={false}
                disabled={false}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": "white",
                      "fontSize": 28,
                    },
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                      },
                      undefined,
                    ],
                    Object {
                      "fontFamily": "Material Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                  ]
                }
              >
                î—’
              </Text>
            </View>
          </View>
        </View>
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Array [
              Object {},
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              Object {
                "color": "#fff",
              },
            ]
          }
        >
          Account
        </Text>
        <View
          style={
            Object {
              "height": 24,
              "width": 24,
            }
          }
        />
      </View>
    </View>
  </View>
  <RCTScrollView
    contentContainerStyle={
      Array [
        Object {
          "flexGrow": 1,
        },
      ]
    }
  >
    <View>
      <View
        style={
          Object {
            "flex": 1,
            "justifyContent": "center",
            "paddingBottom": 84,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#F7F7F7",
                "borderRadius": 4,
                "flex": 1,
                "marginLeft": 10,
                "marginRight": 10,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "marginLeft": 15,
                  "marginRight": 15,
                },
                undefined,
                undefined,
              ]
            }
          >
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                  },
                  undefined,
                ]
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#9E9E9E",
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "padding": 19,
                    },
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    Object {
                      "backgroundColor": "#86b200",
                    },
                    undefined,
                    Object {
                      "padding": 12,
                    },
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                  ]
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  disabled={false}
                  ellipsizeMode="tail"
                  style={
                    Array [
                      Object {},
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      Array [
                        Object {
                          "color": "white",
                          "fontSize": 20,
                        },
                        undefined,
                        Object {
                          "fontSize": 17.5,
                        },
                        undefined,
                        undefined,
                        Object {
                          "fontWeight": "bold",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  Log Out
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
